source/common/http/codec_client.cc

```diff
@@ -56,6 +46,18 @@ CodecClient::CodecClient(Type type, Network::ClientConnectionPtr&& connection,
 
 CodecClient::~CodecClient() = default;
 
+void CodecClient::connect() {
+  // In general, codecs are handed new not-yet-connected connections, but in the
+  // case of ALPN, the codec may be handed an already connected connection.
+  if (!connection_->connecting()) {
+    ASSERT(connection_->state() == Network::Connection::State::Open);
+    connected_ = true;
+  } else {
+    ENVOY_CONN_LOG(debug, "connecting", *connection_);
+    connection_->connect();
+  }
+}
```

Comment:
What can be done to ensure that we don't forget to call connect() after creating a CodecClient ?

One thing that occurs to me: add a bool connect_called_ and verify that it is set on CodecClient destruction. It would detect late, but it would still detect and be less intrusive than having said assert at the top of each CodecClient method.